components/esp_lvgl_port/esp_lvgl_port.c

```diff
@@ -0,0 +1,418 @@
+
+// LVGL callbacks
+static void lvgl_port_flush_callback(lv_disp_drv_t *drv, const lv_area_t *area, lv_color_t *color_map);
+static void lvgl_port_update_callback(lv_disp_drv_t *drv);
+static void lvgl_port_touchpad_read(lv_indev_drv_t *indev_drv, lv_indev_data_t *data);
+static void lvgl_port_pix_callback(lv_disp_drv_t *drv, uint8_t *buf, lv_coord_t buf_w, lv_coord_t x, lv_coord_t y, lv_color_t color, lv_opa_t opa);
+/*******************************************************************************
+* Public API functions
+*******************************************************************************/
+
+esp_err_t lvgl_port_init(void)
+{
+    esp_err_t ret = ESP_OK;
+    memset(&lvgl_port_ctx, 0, sizeof(lvgl_port_ctx));
+
+    /* LVGL init */
+    lv_init();
+    /* Tick init */
+    ESP_RETURN_ON_ERROR(lvgl_port_tick_init(), TAG, "");
+    /* Create task */
+    lvgl_port_ctx.lvgl_mux = xSemaphoreCreateMutex();
+    ESP_GOTO_ON_FALSE(lvgl_port_ctx.lvgl_mux, ESP_ERR_NO_MEM, err, TAG, "Create LVGL mutex fail!");
+    xTaskCreate(lvgl_port_task, "LVGL task", 4096, NULL, CONFIG_LVGL_PORT_TASK_PRIORITY, NULL);
+
+err:
+    if (ret != ESP_OK) {
+        lvgl_port_deinit();
+    }
+
+    return ret;
+}
+
+esp_err_t lvgl_port_deinit(void)
+{
+    /* Stop and delete timer */
+    if (lvgl_port_ctx.tick_timer != NULL) {
+        esp_timer_stop(lvgl_port_ctx.tick_timer);
+        esp_timer_delete(lvgl_port_ctx.tick_timer);
+        lvgl_port_ctx.tick_timer = NULL;
+    }
+
+    /* Stop running task */
+    if (!lvgl_port_ctx.running) {
+        lvgl_port_ctx.running = false;
+    } else {
+        lvgl_port_task_deinit();
+    }
+
+    return ESP_OK;
+}
+
+lv_disp_t *lvgl_port_add_disp(const lvgl_port_display_cfg_t *disp_cfg)
+{
+    esp_err_t ret = ESP_OK;
+    lv_disp_t *disp = NULL;
+    lv_color_t *buf1 = NULL;
+    lv_color_t *buf2 = NULL;
+    assert(disp_cfg != NULL);
+    assert(disp_cfg->handle != NULL);
+    assert(disp_cfg->buffer_size > 0);
+    assert(disp_cfg->hres > 0);
+    assert(disp_cfg->vres > 0);
+
+    /* Display context */
+    lvgl_port_display_ctx_t *disp_ctx = malloc(sizeof(lvgl_port_display_ctx_t));
+    ESP_GOTO_ON_FALSE(disp_ctx, ESP_ERR_NO_MEM, err, TAG, "Not enough memory for display context allocation!");
+    disp_ctx->handle = disp_cfg->handle;
+    disp_ctx->rotation.swap_xy = disp_cfg->rotation.swap_xy;
+    disp_ctx->rotation.mirror_x = disp_cfg->rotation.mirror_x;
+    disp_ctx->rotation.mirror_y = disp_cfg->rotation.mirror_y;
+
+    /* alloc draw buffers used by LVGL */
+    /* it's recommended to choose the size of the draw buffer(s) to be at least 1/10 screen sized */
+    buf1 = heap_caps_malloc(disp_cfg->buffer_size * sizeof(lv_color_t), MALLOC_CAP_DMA);
+    ESP_GOTO_ON_FALSE(buf1, ESP_ERR_NO_MEM, err, TAG, "Not enough memory for LVGL buffer (buf1) allocation!");
+    if (disp_cfg->double_buffer) {
+        buf2 = heap_caps_malloc(disp_cfg->buffer_size * sizeof(lv_color_t), MALLOC_CAP_DMA);
```

Comment:
it will be feasible to also allow the user to provide the draw buffer instead of we allocating for them

Reply:
and the draw buffer is not always needed to be DMA capable

Reply:
I would like to add flags like: `ESP_LVGL_PORT_FLAG_BUFF_DMA`, `ESP_LVGL_PORT_FLAG_BUFF_SPIRAM` and create buffer based these flags.

Reply:
Added.